### ISSUE TITLE
Buildcraft Incompatibility

### ISSUE THREAD

**Issue by Rongmario** [Wabbit0101/mods_hoardercraft#50]
https://paste.dimdev.org/egujilejeq.mccrash

Seems to be on similar lines as the Astral Sorcery crash.

**Comment by Doomgull** [comment]
That's because it's the same base issue as #47 , based on reading that.

**Comment by Wabbit0101** [comment]
Yes. The problem code in the silicon facadestatemanager line 197 :

> Block block = state.getBlock();  //OK
    Item item = Item.getItemFromBlock(block); //OK
        if (item == Items.AIR) {
            item = block.getItemDropped(state, new Random(0), 0);
        }
        return new ItemStack(item, 1, block.damageDropped(state)); //**NOT-OK** (use block.getMetaFromState instead)

The block's item and the dropped-item's meta are not necessarily belonging to the same object (block's item != dropped item). I believe this code should be updated to use Block.getMetaFromState so as to not assume what's dropped is the same as the block's itemblock.

**Comment by AlexIIL** [comment]
`Block.damageDropped` returns a value used for the metadata in items. `Block.getMetaFromState` returns a value only for use in block serialisation in-world, so I'm not sure why it would be better. In addition changing the code to use getMetaFromState breaks blocks like oak wood, which only has a single metadata in item form, but many metadata values in block form. Only one of those item stacks are actually obtainable (as the orientation of the block isn't decided by the itemstack for normal blocks, but it is for facades).

Instead would it be possible for you to do some bounds checking on the given metadata values?

**Comment by Wabbit0101** [comment]
Here's my understanding (long): The bounds checking would prevent a crash but the code is still not correct. Block.damageDropped is the meta data ONLY for the result of Block.getItemDropped. It is not necessarily the meta value that matches the block's ItemBlock _when you have a block that drops a something other than itself that is not air_. 

A simple example might help: I have block "B" that has two states S1 and S2. When in state S1 the block drops ItemA with meta=5 if broken. When in state S2 the block drops ItemB with meta=123. Notice how meta=5 and meta=123 as returned by Block.damageDropped is linked to ItemA or ItemB and not block B's ItemBlock which is what Item.getItemFromBlock returns.  Block B's implementation of getMetaFromState could say if in I'm S1 then return 0 otherwise return 1 to ensure the meta values remain within the valid 0-15 range for blocks.  It's getStateFromMeta would do the reverse looking for meta=0 or 1 to return S1 or S2 property values. If you passed in 5 or 123, it's not expecting that as it's never returned 5 or 123 from the partner API. (The ItemA+ItemB are separate things that the block can't 'force' to use meta=0 or meta=1 to match up...or to stay within the valid  [0-15] range for state meta values.). 

A concrete (contrived) example: Think of a simple Lucky Block that can drop 1 of 15 loot items; it uses index 0-15 to pick the item to drop; the dropped item is stored in a fixed ItemStack array "someplace". It overrides getPickBlock and getItem to make sure both modes work as expected. Meanwhile, damageDropped looks at the incoming state, maps it to an index 0-15 and returns whatever meta value _that loot item_ has. (To prevent crashes if the incoming meta is not 0-15, it returns a lump of coal...no crashing, but not correct either.)

Maybe I have this all wrong...but anyhoos...VFP will be changed to bounds check to prevent crashing but buildcraft and any other mod that assumes ItemBlock meta is-same-as droppedDamage value  be returned the wrong thing (whatever meta 0 state is) for all values. Thank goodness for the upcoming flattening I guess...

**Comment by AlexIIL** [comment]
Yes, we don't use an ideal method. Ideally we would use `Block.getPickBlock`, but that requires both a player and a world, neither of which we can get. (We had issues with sponge when we tried to use a fake world unfortunately).

I've found that the *current* (flawed) method seems to work better than the alternatives with vanilla, and I'm basically just hoping that it sort of works with most mods as well. Failing that we have some API overrides that allow custom mappings from `IBlockState` to `ItemStack` if some-one desperately wants to make it work.
